specification/flag-evaluation/hooks.md

```diff
@@ -2,4 +2,58 @@
 
 ## Overview
 
-// TODO: this is a stub.
+Hooks are a mechanism which can tie into the lifecycle of flag evaluation. They
+operate similarly to middleware in many web frameworks.
+
+### Hook context
+
+Hook context exists to provide hooks with information about the invocation.
+
+> Hook context **MUST** provide: the flag key, flag type, evaluation context, default value and a list of  executed hooks by stage.
+
+> Hook context **SHOULD** provide: provider, client
+
+> flag key, flag type, default value properties **MUST** be immutable.
+
+> The evaluation context **MUST** be mutable only within the `before` hook.
```

Comment:
I like both of these ideas, but I'm not sure all languages have immutability concepts, so you might want to make this conditional to available language mechanisms.